### ISSUE TITLE
[SHIELD] Fix constrained_decoding.py NotImplementedError stub

### ISSUE THREAD

**Issue by andrewlasiter** [andrewlasiter/FOIA-Madlibs#200]
**Estimate**: 2 days | **Component**: Constrained Beam Search

`filter_logits_by_vocabulary()` in `scripts/constrained_decoding.py:582` raises `NotImplementedError`. Any infill call that uses vocabulary-filtered beam search crashes at runtime.

## Problem

```python
def filter_logits_by_vocabulary(logits, vocabulary_set):
    raise NotImplementedError("Vocabulary filtering not yet implemented")
```

## Tasks

- [ ] Implement logit filtering: zero out logits for tokens not in `vocabulary_set`
- [ ] Handle tokenizer edge cases (subword tokens that form valid words)
- [ ] Add unit tests covering: empty vocabulary, full vocabulary, partial match
- [ ] Verify `ConstrainedBeamSearch` integration works end-to-end

## Files

* `scripts/constrained_decoding.py` (modify `filter_logits_by_vocabulary`)
* `tests/test_constrained_decoding.py` (add vocabulary filter tests)

## Acceptance Criteria

* `filter_logits_by_vocabulary()` no longer raises `NotImplementedError`
* Filtered logits only allow tokens in the vocabulary set
* Existing constrained decoding tests still pass
